executor/executor.go

```diff
@@ -2348,3 +2362,457 @@ func isWeakConsistencyRead(ctx sessionctx.Context, node ast.Node) bool {
+	}
+	return checksums, nil
+}
+
+// HandleTask implements the Worker interface.
+func (w *checkIndexWorker) HandleTask(task checkIndexTask) {
+	defer w.e.wg.Done()
+	idxInfo := w.indexInfos[task.indexOffset]
+	bucketSize := int(CheckTableFastBucketSize.Load())
+
+	ctx := kv.WithInternalSourceType(w.e.contextCtx, kv.InternalTxnAdmin)
+
+	trySaveErr := func(err error) {
+		w.e.err.CompareAndSwap(nil, &err)
+	}
+
+	se, err := w.e.base().getSysSession()
+	if err != nil {
+		trySaveErr(err)
+		return
+	}
+	se.GetSessionVars().OptimizerUseInvisibleIndexes = true
+	defer func() {
+		se.GetSessionVars().OptimizerUseInvisibleIndexes = false
+		w.e.base().releaseSysSession(ctx, se)
+	}()
+
+	var pkCols []string
+	var pkTypes []*types.FieldType
+	switch {
+	case w.e.table.Meta().IsCommonHandle:
+		pkColsInfo := w.e.table.Meta().GetPrimaryKey().Columns
+		for _, colInfo := range pkColsInfo {
+			colStr := colInfo.Name.O
+			pkCols = append(pkCols, colStr)
+		}
+	case w.e.table.Meta().PKIsHandle:
+		pkCols = append(pkCols, w.e.table.Meta().GetPkName().O)
+	default: // support decoding _tidb_rowid.
+		pkCols = append(pkCols, model.ExtraHandleName.O)
+	}
+
+	// CheckSum of (handle + index columns).
+	var md5HandleAndIndexCol strings.Builder
+	md5HandleAndIndexCol.WriteString("crc32(md5(concat_ws(0x2, ")
+	for _, col := range pkCols {
+		md5HandleAndIndexCol.WriteString(dbutil.ColumnName(col))
+		md5HandleAndIndexCol.WriteString(", ")
+	}
+	for offset, col := range idxInfo.Columns {
+		md5HandleAndIndexCol.WriteString("ifnull(")
+		tblCol := w.table.Meta().Columns[col.Offset]
+		if tblCol.IsGenerated() && !tblCol.GeneratedStored {
+			md5HandleAndIndexCol.WriteString(tblCol.GeneratedExprString)
+		} else {
+			md5HandleAndIndexCol.WriteString(dbutil.ColumnName(col.Name.O))
+		}
+		md5HandleAndIndexCol.WriteString(", 0x1)")
+		if offset != len(idxInfo.Columns)-1 {
+			md5HandleAndIndexCol.WriteString(", ")
+		}
+	}
+	md5HandleAndIndexCol.WriteString(")))")
+
+	// Used to group by and order.
+	var md5Handle strings.Builder
+	md5Handle.WriteString("crc32(md5(concat_ws(0x2, ")
+	for i, col := range pkCols {
+		md5Handle.WriteString(dbutil.ColumnName(col))
+		if i != len(pkCols)-1 {
+			md5Handle.WriteString(", ")
+		}
+	}
+	md5Handle.WriteString(")))")
+
+	handleColumnField := strings.Join(pkCols, ", ")
+	var indexColumnField strings.Builder
+	for offset, col := range idxInfo.Columns {
+		indexColumnField.WriteString(dbutil.ColumnName(col.Name.O))
+		if offset != len(idxInfo.Columns)-1 {
+			indexColumnField.WriteString(", ")
+		}
+	}
+
+	tableRowCntToCheck := int64(0)
+
+	offset := 0
+	mod := 1
+	meetError := false
+
+	lookupCheckThreshold := int64(100)
+	checkOnce := false
+
+	if w.e.ctx.GetSessionVars().SnapshotTS != 0 {
+		se.GetSessionVars().SnapshotTS = w.e.ctx.GetSessionVars().SnapshotTS
+		defer func() {
+			se.GetSessionVars().SnapshotTS = 0
+		}()
+	}
+	_, err = se.(sqlexec.SQLExecutor).ExecuteInternal(ctx, "begin")
```

Comment:
It seem that I don't see rollback or commit for this txn. Am I wrong?

Reply:
rollback in releaseSysSession